hl.vds.new_combiner fails at compile with 'IR already in memo' when a combine job writes >=3 datasets (write_variant_datasets)

### What happened?

Running `hl.vds.new_combiner(...).run()` on a moderately large cohort fails at **driver compile time** (before any distributed work) with:

```
is.hail.utils.HailException: error trying to rebuild IR:
Caused by: java.lang.RuntimeException: IR already in memo:
  NA(struct{oldCtx: tuple(int32, interval<struct{locus: locus<GRCh38>}>,
     struct{contig: str, start: int32, end: int32}), writeCtx: str})
```

Python traceback (top frames):

```
File ".../hail/vds/combiner/variant_dataset_combiner.py", line 358, in run
File ".../hail/vds/combiner/variant_dataset_combiner.py", line 430, in step
File ".../hail/vds/combiner/variant_dataset_combiner.py", line 629, in _step_gvcfs
File ".../hail/vds/methods.py", line 16, in write_variant_datasets
...
hail.utils.java.FatalError: RuntimeException: IR already in memo: NA(struct{...writeCtx: str})
```

The failing lowering pass is `Optimize: compileLowerer` / `LowerOrInterpretNonCompilable`; the exception is thrown from `RefEquality.scala:29` (`throw new RuntimeException(s"IR already in memo: ...")`).

### Root cause (analysis)

`_step_gvcfs` ends a combine job with `hl.vds.write_variant_datasets(merge_vds, paths, ...)` (`variant_dataset_combiner.py:629`), which writes **multiple `VariantDataset`s in a single IR**. When that IR bundles enough per-dataset `WriteMetadata`/`NA` nodes, the same IR node **object** appears more than once and `RefEquality`'s `Memo` rejects the duplicate. A job producing exactly **one** dataset takes the single-write `_write_final` path (`variant_dataset_combiner.py:618-620`) and compiles fine.

`stderr` shows the duplicated nodes directly — e.g. `!50`/`!51` and `!63`/`!65` are identical `NA[Struct{oldCtx:…, writeCtx}]` nodes.

This looks like a not-yet-covered case of the ongoing no-sharing campaign (#15449, #15454, #15458, #15462, #15499, #15500, #15513, #15520, #15525): none of those touch the combiner's `write_variant_datasets` / multi-`MatrixWrite` path.

### Minimal reproduction (local mode, no cloud, ~1 min)

Uses Hail's own bundled chr22 test gVCFs (`hail/test/resources/gvcfs/1kg_chr22/`). **The threshold is exact: a combine job writing 2 datasets compiles OK; ≥3 datasets fail.**

```python
import glob, os, tempfile, hail as hl

gvcfs = sorted(glob.glob('hail/test/resources/gvcfs/1kg_chr22/*.hg38.g.vcf.gz'))[:6]
hl.init(master='local[2]')
tmp = tempfile.mkdtemp()

# branch_factor=2, gvcf_batch_size=3, 6 gVCFs -> ceil(6/2)=3 datasets
# bundled into one write_variant_datasets IR -> RuntimeException: IR already in memo
hl.vds.new_combiner(
    output_path=os.path.join(tmp, 'repro.vds'),
    temp_path=tmp,
    gvcf_paths=gvcfs,
    reference_genome='GRCh38',
    use_exome_default_intervals=True,
    branch_factor=2,
    gvcf_batch_size=3,
).run()
```

Bisected results (`branch_factor=2`, `n_gvcfs = 2 * gvcf_batch_size`):

| datasets in one write | result |
|---|---|
| 1 (default params, small n) | OK |
| 2 (`gvcf_batch_size=2`, n=4) | OK |
| **3** (`gvcf_batch_size=3`, n=6) | **FAIL** |
| 4, 5, 12, 25 | FAIL |

This also explains why default-parameter runs only fail past a certain cohort size: with default `branch_factor=100`, `gvcf_batch_size=50`, a run of 250 gVCFs gives `ceil(250/100) = 3` datasets in the first job → fail, whereas ≤ ~200 gVCFs give ≤ 2 datasets → OK.

`gvcf_batch_size` and `import_interval_size` do **not** avoid it — the trigger is the number of datasets in one write, not partitioning.

### Workaround

Combine in sub-cohorts (each `new_combiner` run kept to a single output dataset), then merge the resulting VDSes with `new_combiner(vds_paths=[...])` — the `_step_vdses` merge path uses single writes. Verified end-to-end on 250 gVCFs.

### Version

Hail 0.2.138 (tree up to date with `main` past the newest merged no-sharing PR #15525).

Reproduced in **two** independent environments (so not a Java/Spark-build artifact):
- Local mode: PySpark 3.5.9, Java 17.
- AWS EMR 7.3.0: Spark 3.5.1-amzn-1, Python 3.11, on a real 250-sample DRAGEN gVCF cohort (1000 Genomes DRAGEN reanalysis).

--- a/hail/hail/src/is/hail/expr/ir/MatrixWriter.scala
+++ b/hail/hail/src/is/hail/expr/ir/MatrixWriter.scala
@@ -2497,17 +2497,16 @@ case class MatrixNativeMultiWriter(
 
     val contextUnionType = TStruct("__matrix_id" -> TInt32, "__options" -> unionType)
 
-    val emptyUnionIRs: IndexedSeq[(Int, IR)] =
-      ArraySeq.tabulate(unionType.size)(i => i -> NA(unionType.types(i)))
-
     val concatenatedContexts =
       flatten(
         MakeArray(
           components.zipWithIndex.map { case (c, matrixId) =>
-            ToArray(mapIR(c.stage.contexts) { ctx =>
+            ToArray(c.stage.contexts.streamMap { ctx =>
               makestruct(
                 "__matrix_id" -> I32(matrixId),
-                "__options" -> MakeTuple(emptyUnionIRs.updated(matrixId, matrixId -> ctx.ir)),
+                "__options" -> MakeTuple(ArraySeq.tabulate(unionType.size) { i =>
+                  i -> (if (i == matrixId) ctx.ir else NA(unionType.types(i)))
+                }),
               )
             })
           },
@@ -2531,25 +2530,22 @@ case class MatrixNativeMultiWriter(
 
         result <- cdaIR(concatenatedContexts, allBroadcasts, "matrix_multi_writer") {
           (ctx, globals) =>
-            Switch(
-              GetField(ctx, "__matrix_id"),
-              default = Die("MatrixId exceeds matrix count", components.head.writePartitionType),
-              cases = components.zipWithIndex.map { case (component, i) =>
-                val binds = component.stage.broadcastVals.map { case (name, _) =>
-                  name -> GetField(globals, name.str)
-                }
-
-                Let(
-                  binds,
+            ctx.get("__options").bind { options =>
+              Switch(
+                ctx.get("__matrix_id"),
+                default = Die("MatrixId exceeds matrix count", components.head.writePartitionType),
+                cases = components.zipWithIndex.map { case (component, i) =>
                   IRBuilder.scoped { b =>
-                    val options = GetField(ctx, "__options")
-                    val ctxRef = b.memoize(GetTupleElement(options, i))
+                    for ((name, _) <- component.stage.broadcastVals)
+                      b.strictMemoize(globals.get(name.str), name)
+
+                    val ctxRef = b.memoize(options.get(i))
                     val rows = b.memoize(component.stage.partition(ctxRef))
                     component.writePartition(rows, ctxRef)
-                  },
-                )
-              },
-            )
+                  }
+                },
+              )
+            }
         }
 
         partCounts = components.map(_.stage.numPartitions).scanLeft(0)(_ + _)

--- a/hail/python/test/hail/matrixtable/test_matrix_table.py
+++ b/hail/python/test/hail/matrixtable/test_matrix_table.py
@@ -1684,16 +1684,13 @@ class Tests(unittest.TestCase):
         assert localized._same(t)
 
     def test_multi_write(self):
-        mt = self.get_mt()
-        f = new_temp_file()
-        hl.experimental.write_matrix_tables([mt, mt], f)
-        path1 = f + '0.mt'
-        path2 = f + '1.mt'
-        mt1 = hl.read_matrix_table(path1)
-        mt2 = hl.read_matrix_table(path2)
-        self.assertTrue(mt._same(mt1))
-        self.assertTrue(mt._same(mt2))
-        self.assertTrue(mt1._same(mt2))
+        mt = hl.utils.range_matrix_table(10, 10)
+        mts = [mt] * 3  # RR: 15604
+        prefix = new_temp_file()
+        paths = hl.experimental.write_matrix_tables(mts, prefix)
+        for path in paths:
+            mt_n = hl.read_matrix_table(path)
+            self.assertTrue(mt._same(mt_n))
 
     def test_matrix_type_equality(self):
         mt = hl.utils.range_matrix_table(1, 1)

